src/soc/rockchip/rk3399.c

```diff
@@ -0,0 +1,377 @@
+    {"GPIO1_B2", 1, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 10, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+
+    {"GPIO2_A0", 2, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 0, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO2_A1", 2, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 1, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO2_B2", 2, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 10, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO2_B1", 2, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 9, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+
+    {"GPIO3_C0", 3, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 16, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO2_B4", 2, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 12, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO4_A5", 4, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 5, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO4_A4", 4, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 4, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+
+    {"GPIO4_D6", 4, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 30, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO4_A6", 4, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 6, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+    {"GPIO4_A7", 4, {GPIO_SWPORTA_DDR}, {GPIO_SWPORTA_DR}, 7, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+};
+
+static int setPin(int pin, int arr[]) {
+    int i;
+    for (i = 0; i < PIN_NUMS; i++) {
+        if (pin == arr[i]) {
+            return i;
+        }
+    }
+}
+
+static int rk3399Setup(void) {
+    if((rk3399->fd = open("/dev/mem", O_RDWR | O_SYNC)) < 0) {
+        wiringXLog(LOG_ERR, "wiringX failed to open /dev/mem for raw memory access");
+        return -1;
+    }
+    if((rk3399->gpio[1] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[1])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[2] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[2])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[3] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[3])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[4] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[4])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    return 0;
+}
+
+static char *rk3399GetPinName(int pin) {
+    return rk3399->layout[pin].name;
```

Comment:
Please add additional validation to check if the pin is inside the allowed pin range.